os/drivers/audio/alc1019.c

```diff
@@ -0,0 +1,1083 @@
+	
+	//alc1019_set_i2s_samplerate(priv);
+	//alc1019_set_i2s_datawidth(priv);
+
+	return;
+}
+
+/****************************************************************************
+ * Name: alc1019_hw_reset
+ *
+ * Description:
+ *   Reset and re-initialize the ALC1019
+ *
+ * Input Parameters:
+ *   priv - A reference to the driver state structure
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+static void alc1019_hw_reset(FAR struct alc1019_dev_s *priv)
+{
+	if (!priv) {
+		return;
+	}
+
+	struct alc1019_lower_s *lower = priv->lower;
+
+	if (lower->control_hw_reset) {
+		lower->control_hw_reset(true);
+		up_mdelay(100);
+		lower->control_hw_reset(false);
+		up_mdelay(100);
+	}
+	alc1019_reconfigure(priv);
+	alc1019_exec_i2c_script(priv, codec_init_mute_on_script, sizeof(codec_init_mute_on_script) / sizeof(t_codec_init_script_entry));
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: alc1019_initialize
+ *
+ * Description:
+ *   Initialize the ALC1019 device.
+ *
+ * Input Parameters:
+ *   i2c     - An I2C driver instance
+ *   i2s     - An I2S driver instance
+ *   lower   - Persistent board configuration data
+ *
+ * Returned Value:
+ *   A new lower half audio interface for the ALC1019 device is returned on
+ *   success; NULL is returned on failure.
+ *
+ ****************************************************************************/
+FAR struct audio_lowerhalf_s *alc1019_initialize(FAR struct i2c_dev_s *i2c, FAR struct i2s_dev_s *i2s, FAR struct alc1019_lower_s *lower)
+{
+
+	FAR struct alc1019_dev_s *priv;
+	uint8_t regval;
+	/* Sanity check */
+	DEBUGASSERT(i2c && lower && i2s);
+
+	auddbg("I2s dev addr is 0x%x\n", i2s);
+
+	/* Allocate a ALC1019 device structure */
+	priv = (FAR struct alc1019_dev_s *)kmm_zalloc(sizeof(struct alc1019_dev_s));
+
+	if (!priv) {
+		return NULL;
+	}
+
+	/* Initialize the ALC1019 device structure.  Since we used kmm_zalloc,
+	 * only the non-zero elements of the structure need to be initialized.
+	 */
+
+	priv->dev.ops = &g_audioops;
+	priv->lower = lower;
+	priv->i2c = i2c;
+	priv->i2s = i2s;
+
+	sem_init(&priv->devsem, 0, 1);
+	sq_init(&priv->pendq);
+
+	/* Software reset.  This puts all ALC1019 registers back in their
+	 * default state.
+	 */
+
+	/*reconfigure the ALC1019 hardwaqre */
+	alc1019_reconfigure(priv);
+	alc1019_exec_i2c_script(priv, codec_init_mute_on_script, sizeof(codec_init_mute_on_script) / sizeof(t_codec_init_script_entry));
+
+#ifndef CONFIG_AUDIO_EXCLUDE_VOLUME
+	priv->volume = ALC1019_SPK_VOL_DEF;
+	alc1019_setvolume(priv);
+#endif
+
+	return &priv->dev;
+
+errout_with_dev:
+	sem_destroy(&priv->devsem);
+	kmm_free(priv);
+	return NULL;
```

Comment:
1. Looks unnecessary
2. If we remove this, there is no free for priv. Is it ok?

Reply:
> 1. Looks unnecessary
> 2. If we remove this, there is no free for priv. Is it ok?
priv is private data of iNode, it should not be free.